### PR TITLE
Robust tests

### PR DESCRIPTION
**Pull request by d-atkins**
#### Adds edge case testing to all tests
- adds messages and ciphertexts containing symbols and punctuation
- adds uppercase letters to `encoder_test`
#### Fixes bug in `Cracker`
- fixes bug that allowed negative root code to be made in rare instances

### CODE REVIEW

File: lib/cracker.rb [hunk 1]
```diff
@@ -11,7 +11,10 @@ def reverse_shifts(ciphertext, known_end = " end")
   end
 
   def root_codes(shifts, offsets)
-    shifts.zip(offsets).map {|shift, offset| shift - offset.to_i}
+    shifts.zip(offsets).map do |shift, offset|
+      diff = shift - offset.to_i
+      diff > 0 ? diff : full_shift(diff)
```
Comment (d-atkins) [inline lib/cracker.rb:16]: found this bug after adding the _very last_ extra test

`Testing: It's Worth It™`
